File: backend/internal/domain/card_talent.go
package domain

import "strconv"

// CardTalent identifies a talent tag (persisted as smallint values in cards.talents).
// Values must stay in sync with frontend `src/constants/cardTalent.js`.
type CardTalent int16

const (
	CardTalentDraconic  CardTalent = 0
	CardTalentEarth     CardTalent = 1
	CardTalentElemental CardTalent = 2
	CardTalentIce       CardTalent = 3
	CardTalentLight     CardTalent = 4
	CardTalentLightning CardTalent = 5
	CardTalentRoyal     CardTalent = 6
	CardTalentShadow    CardTalent = 7
)

var cardTalentNames = map[CardTalent]string{
	CardTalentDraconic:  "Draconic",
	CardTalentEarth:     "Earth",
	CardTalentElemental: "Elemental",
	CardTalentIce:       "Ice",
	CardTalentLight:     "Light",
	CardTalentLightning: "Lightning",
	CardTalentRoyal:     "Royal",
	CardTalentShadow:    "Shadow",
}

// Valid reports whether t is a defined CardTalent constant.
func (t CardTalent) Valid() bool {
	_, ok := cardTalentNames[t]
	return ok
}

// String returns the wire/display name or "CardTalent("+decimal+")" if unknown.
func (t CardTalent) String() string {
	if name, ok := cardTalentNames[t]; ok {
		return name
	}
	return "CardTalent(" + strconv.FormatInt(int64(t), 10) + ")"
}

// CardTalents returns every defined talent in ascending ID order.
func CardTalents() []CardTalent {
	return []CardTalent{
		CardTalentDraconic,
		CardTalentEarth,
		CardTalentElemental,
		CardTalentIce,
		CardTalentLight,
		CardTalentLightning,
		CardTalentRoyal,
		CardTalentShadow,
	}
}

File: src/constants/cardTalent.js
/**
 * Card talent enum IDs (smallint in `talents` arrays). Must match `backend/internal/domain/card_talent.go`.
 * @readonly
 */
export const CardTalent = Object.freeze({
  Draconic: 0,
  Earth: 1,
  Elemental: 2,
  Ice: 3,
  Light: 4,
  Lightning: 5,
  Royal: 6,
  Shadow: 7,
});

/** Ordered names for each ID 0..7. */
export const CARD_TALENT_NAMES = Object.freeze([
  "Draconic",
  "Earth",
  "Elemental",
  "Ice",
  "Light",
  "Lightning",
  "Royal",
  "Shadow",
]);

/**
 * @param {number} id
 * @returns {string | undefined}
 */
export function cardTalentName(id) {
  return CARD_TALENT_NAMES[id];
}

/**
 * @param {number} id
 * @returns {boolean}
 */
export function isValidCardTalentId(id) {
  return Number.isInteger(id) && id >= 0 && id < CARD_TALENT_NAMES.length;
}
